clippy_lints/src/loops.rs

```diff
@@ -1953,6 +1987,63 @@ fn check_for_single_element_loop<'tcx>(
     }
 }
 
+/// Check for unnecessary `if let` usage in a for loop where only the `Some` or `Ok` variant of the
+/// iterator element is used.
+fn check_manual_flatten<'tcx>(
+    cx: &LateContext<'tcx>,
+    span: Span,
+    pat: &'tcx Pat<'_>,
+    arg: &'tcx Expr<'_>,
+    body: &'tcx Expr<'_>,
+) {
+    if_chain! {
+        // Ensure the `if let` statement is the only expression in the for-loop
+        if let ExprKind::Block(ref block, _) = body.kind;
+        if block.stmts.is_empty();
+        if let Some(inner_expr) = block.expr;
+        if let ExprKind::Match(
+            ref match_expr, ref match_arms, MatchSource::IfLetDesugar{ contains_else_clause }
+        ) = inner_expr.kind;
+        if !contains_else_clause;
+        // Ensure match_expr in `if let` statement is the same as the pat from the for-loop
+        if snippet(cx, match_expr.span, "..") == snippet(cx, pat.span, "..");
```

Comment:
@camsteffen After failing some dogfood test, I realized that the `match_expr` and `pat` must be the same for this lint to trigger:
```rust
for pat in iter {
    if let Some(n) = match_expr {
        ... ...
    }
}
```
Otherwise the lint must handle even more complicated scenarios. However, I could only come up with a hacky approach to match the `snippet`s. Do you know how we could compare a `Pat` and an `Expr` to ensure they refer to the same thing?

Reply:
First check that `pat` is a `PatKind::Binding` and get the `HirId`. Then check that `match_expr` is a `ExprKind::Path` that contains a `QPath::Resolved` with a `Res::Local` that contains the same `HirId` as the one in the `PatKind::Binding`.

Reply:
Thanks, this helps a lot! I saw the matching `HirId` but was not sure if that was the thing to match for!